### DISCUSSION TITLE
Channel has enough Balance, but INSUFFICIENT_BALANCE Errors for HTLC Forwards are happening

### DISCUSSION THREAD

**Discussion by icebaker** [lightningnetwork/lnd#7383] [Troubleshooting]
I'm monitoring events through [router/SubscribeHtlcEvents](https://github.com/lightningnetwork/lnd/blob/master/lnrpc/routerrpc/router.proto#L113).

I started receiving `INSUFFICIENT_BALANCE` forwarding errors for a Channel that actually appears to have more than enough balance:

| Description              | mSats         | Sats      |
|                     ---: |          ---: |      ---: |
| `local_balance`          | 1,137,953,000 | 1,137,953 |
| `outgoingAmtMsat`        |   200,004,802 |   200,005 |
| `local_chan_reserve_sat` |    60,878,000 |    60,878 |
| `dust_limit_sat`         |       354,000 |       354 |

I was able to forward `50` payments through the channel successfully before, and a day ago, I started to have this issue after a rebalance - with feels like just a coincidence and not related to the rebalance.

When the failure happens, I receive two failures in a row (maybe a kind of retry is happening?).

Example of a failure (the two in a row, as I mentioned):
```json
{
  "eventType": "FORWARD",
  "timestampNs": "1675520061868577498",
  "linkFailEvent": {
    "info": {
      "incomingAmtMsat": "200026203",
      "outgoingAmtMsat": "200004802",
      "incomingTimelock": 775221,
      "outgoingTimelock": 775181
    },
    "wireFailure": "TEMPORARY_CHANNEL_FAILURE",
    "failureDetail": "INSUFFICIENT_BALANCE",
    "failureString": "insufficient bandwidth to route htlc"
  },
  "incomingHtlcId": "13726",
  "incomingChannelId": "850099509773795329",
  "outgoingChannelId": "850181973150531585"
}
```

```json
{
  "eventType": "FORWARD",
  "timestampNs": "1675520061863558241",
  "linkFailEvent": {
    "info": {
      "incomingAmtMsat": "200044405",
      "outgoingAmtMsat": "200023002",
      "incomingTimelock": 775221,
      "outgoingTimelock": 775181
    },
    "wireFailure": "TEMPORARY_CHANNEL_FAILURE",
    "failureDetail": "INSUFFICIENT_BALANCE",
    "failureString": "insufficient bandwidth to route htlc"
  },
  "incomingHtlcId": "13725",
  "incomingChannelId": "850099509773795329",
  "outgoingChannelId": "850181973150531585"
}
```

Outgoing Channel State

`lightning/list-channels`
```json
{
  "active": true,
  "remote_pubkey": "0298f6074a454a1f5345cb2a7c6f9fce206cd0bf675d177cdbf0ca7508dd28852f",
  "channel_point": "260e87eb5a49498291991311628ba8faed8503a0d88a43b208286b4d3738ca0e:1",
  "chan_id": 850181973150531585,
  "capacity": 6087836,
  "local_balance": 1137953,
  "remote_balance": 4946763,
  "commit_fee": 3120,
  "commit_weight": 724,
  "fee_per_kw": 4309,
  "unsettled_balance": 0,
  "total_satoshis_sent": 17172941,
  "total_satoshis_received": 12226178,
  "num_updates": 14914,
  "pending_htlcs": [
    
  ],
  "csv_delay": 432,
  "private": false,
  "initiator": true,
  "chan_status_flags": "ChanStatusDefault",
  "local_chan_reserve_sat": 60878,
  "remote_chan_reserve_sat": 60878,
  "static_remote_key": true,
  "commitment_type": "STATIC_REMOTE_KEY",
  "lifetime": 747507,
  "uptime": 746788,
  "close_address": "",
  "push_amount_sat": 0,
  "thaw_height": 0,
  "local_constraints": {
    "csv_delay": 432,
    "chan_reserve_sat": 60878,
    "dust_limit_sat": 354,
    "max_pending_amt_msat": 18446744073709551615,
    "min_htlc_msat": 0,
    "max_accepted_htlcs": 5
  },
  "remote_constraints": {
    "csv_delay": 731,
    "chan_reserve_sat": 60878,
    "dust_limit_sat": 546,
    "max_pending_amt_msat": 6026958000,
    "min_htlc_msat": 1,
    "max_accepted_htlcs": 483
  },
  "alias_scids": [
    
  ],
  "zero_conf": false,
  "zero_conf_confirmed_scid": 0
}
```

`lightning/get-chan-info`
```json
{
  "channel_id": 850181973150531585,
  "chan_point": "260e87eb5a49498291991311628ba8faed8503a0d88a43b208286b4d3738ca0e:1",
  "last_update": 1675507609,
  "node1_pub": "0298f6074a454a1f5345cb2a7c6f9fce206cd0bf675d177cdbf0ca7508dd28852f",
  "node2_pub": "02d3c80335a8ccb2ed364c06875f32240f36f7edb37d80f8dbe321b4c364b6e997",
  "capacity": 6087836,
  "node1_policy": {
    "time_lock_delta": 34,
    "min_htlc": 1,
    "fee_base_msat": 0,
    "fee_rate_milli_msat": 3,
    "disabled": false,
    "max_htlc_msat": 6026958000,
    "last_update": 1675391942,
    "custom_records": {
      
    }
  },
  "node2_policy": {
    "time_lock_delta": 40,
    "min_htlc": 1000,
    "fee_base_msat": 0,
    "fee_rate_milli_msat": 107,
    "disabled": false,
    "max_htlc_msat": 6045000000,
    "last_update": 1675507609,
    "custom_records": {
      
    }
  },
  "custom_records": {
    
  }
}
```

I was suspicious about this being exactly [this issue](https://github.com/lightningnetwork/lnd/issues/5468) because `max_pending_amt_msat`=`18446744073709551615` and `min_htlc_msat`=`0`:

```json
"local_constraints": {
  "csv_delay": 432,
  "chan_reserve_sat": 60878,
  "dust_limit_sat": 354,
  "max_pending_amt_msat": 18446744073709551615,
  "min_htlc_msat": 0,
  "max_accepted_htlcs": 5
},
```

Plus, the Node to who I have the channel opened [apparently](https://1ml.com/node/0298f6074a454a1f5345cb2a7c6f9fce206cd0bf675d177cdbf0ca7508dd28852f) runs on [CLN](https://github.com/ElementsProject/lightning):

```txt
Owner Info
CLN 0.11.2-gentoo-r113
```

But, following the issue, appears to be something already fixed by [#5478](https://github.com/lightningnetwork/lnd/pull/5478).

From my math, there should be no reason for an `INSUFFICIENT_BALANCE` error, but I keep receiving HTLC forwarding failures for this channel.

Thoughts about what could be happening and suggestions about how to debug this?

*Environment*

- LND `0.15.5-beta commit=v0.15.5-beta` `c0a09209782b1c62c3393fcea0844e095c25046b`
- `Linux 5.15.0-58-generic #64-Ubuntu SMP x86_64 x86_64 x86_64 GNU/Linux`
- Bitcoin Core `24.0.1`

**Comment by bitromortac** [comment]
Thank you for the detailed report. I'm looking into this. For completeness, do you run any other tools that intercept HTLCs?

https://github.com/lightningnetwork/lnd/issues/7108 may be related, although both balances are high (so I'd rule this out). The number of accepted HTLCs of your peer is also very low `"max_accepted_htlcs": 5`. It could be that those slots were already occupied at that time.

Could you send to yourself to trigger a insufficient balance error (dest should be your node pubkey, this should not settle if you don't have keysend activated)?:
`lncli sendpayment --keysend -d 02d3c80335a8ccb2ed364c06875f32240f36f7edb37d80f8dbe321b4c364b6e997 --allow_self_payment -a 10000 --outgoing_chan_id 850181973150531585`

**Comment by icebaker** [comment]
> _do you run any other tools that intercept HTLCs_

Yes, I do run other tools, but I'm not sure if they actually intercept HTLCs. My guess would be that they don't. The tools:

- [Umbrel](https://github.com/getumbrel/umbrel)
- [LNDg](https://github.com/cryptosharks131/lndg)
- [Mempool](https://github.com/mempool/mempool)
- [Ride The Lightning](https://github.com/Ride-The-Lightning/RTL)
- [ThunderHub](https://github.com/apotdevin/thunderhub)
- [Lightning Terminal](https://github.com/lightninglabs/lightning-terminal)
- [Lightning Shell](https://github.com/ibz/lightning-shell)

I tried some autopilot features from _LNDg_ a while ago, which may have changed some things, like the `max_accepted_htlcs`... But this is a guess; no evidence.

> _Could you send to yourself to trigger a insufficient balance error_

No errors, worked as expected:

```bash
lncli sendpayment \
  --keysend -d 02d3c80335a8ccb2ed364c06875f32240f36f7edb37d80f8dbe321b4c364b6e997 \
  --allow_self_payment \
  -a 10000 \
  --outgoing_chan_id 850181973150531585
```

```text
+------------+--------------+--------------+--------------+------+----------+--------------------+------------------------------------+
| HTLC_STATE | ATTEMPT_TIME | RESOLVE_TIME | RECEIVER_AMT | FEE  | TIMELOCK | CHAN_OUT           | ROUTE                              |
+------------+--------------+--------------+--------------+------+----------+--------------------+------------------------------------+
| SUCCEEDED  |        0.037 |        5.175 | 10000        | 0.07 |   775644 | 850181973150531585 | BCash_Is_Trash->icebaker/old-stone |
+------------+--------------+--------------+--------------+------+----------+--------------------+------------------------------------+
Amount + fee:   10000 + 0.07 sat
Payment hash:   402be5ba65c4f28288ae52f83a929542224266118cecc182b2cc139249107f0a
Payment status: SUCCEEDED
```

> _It could be that those slots were already occupied at that time._

Sounds reasonable.

An update: Channel got "back to normal," and now I'm routing payments.

So, here's what I'm going to do:

I will remove all those tools. I was experimenting, but now I have my own tooling and don't need them anymore, so I believe it would be easier to debug things without potential interference/noise from them.

Additionally, sounds like, for debugging purposes, it would be very helpful to have a "snapshot" of the node/channels at the exact moment of the failure. Gonna implemented this to evaluate if it will be useful in potential future failures.

About `max_accepted_htlcs`: Do we have some documentation or suggestions about how to define a good value here? What heuristics would be applicable?

Also, how would it be possible, at the moment of the failure, to evaluate if this (`max_accepted_htlcs`) is the potential cause of the error? Perhaps checking the `PendingHtlcs` [gRPC call](https://github.com/lightningnetwork/lnd/blob/master/lnrpc/lightning.pb.go#L4511)?

Let me know if you need additional info, and thanks for the help!
